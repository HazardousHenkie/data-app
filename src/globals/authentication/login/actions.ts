import { action } from 'typesafe-actions'

import AuthenticationState from '../types'
import ActionTypes from './constants'

export const loginRequest = (loginToken: string) =>
    action(ActionTypes.LOGIN_REQUEST, loginToken)
export const loginSuccess = (user: AuthenticationState['user']) =>
    action(ActionTypes.LOGIN_SUCCESS, user)
export const loginError = (error: Error | boolean) =>
    action(ActionTypes.LOGIN_ERROR, error)
