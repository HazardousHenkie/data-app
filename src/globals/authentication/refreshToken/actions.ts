import { action } from 'typesafe-actions'

import AuthenticationState from '../types'
import ActionTypes from './constants'

export const getRefreshTokenRequest = (id: string) =>
    action(ActionTypes.GET_REFRESH_TOKEN_REQUEST, id)
export const getRefreshTokenSuccess = (user: AuthenticationState['user']) =>
    action(ActionTypes.GET_REFRESH_TOKEN_SUCCESS, user)
export const getRefreshTokenError = (error: Error | boolean) =>
    action(ActionTypes.GET_REFRESH_TOKEN_ERROR, error)
