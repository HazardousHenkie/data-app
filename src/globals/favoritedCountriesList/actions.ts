import { action } from 'typesafe-actions'

import { ResponseError } from 'utils/request'
import ActionTypes from './constants'

import FavoritedCountriesState from './types'

export const getFavoritedCountries = () =>
    action(ActionTypes.GET_FAVORITED_COUNTRIES)

export const setFavoritedCountries = (
    favoritedCountries: FavoritedCountriesState['countries']
) => action(ActionTypes.SET_FAVORITED_COUNTRIES, favoritedCountries)

export const getFavoritedCountriesError = (error: ResponseError) =>
    action(ActionTypes.GET_FAVORITED_COUNTRIES_ERROR, error)
