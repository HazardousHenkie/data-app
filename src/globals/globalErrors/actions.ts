import { action } from 'typesafe-actions'

import ActionTypes from './constants'

export const getErrors = () => action(ActionTypes.GET_ERRORS)

export const setError = (error: Error) => action(ActionTypes.SET_ERROR, error)

export const removeError = () => action(ActionTypes.REMOVE_ERROR)

export const setErrors = (errors: Error[]) =>
    action(ActionTypes.SET_ERRORS, errors)
