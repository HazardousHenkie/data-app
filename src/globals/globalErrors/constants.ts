import ErrorsState from './types'

enum ActionTypes {
    GET_ERRORS = 'components/globalErrors/GET_ERRORS',
    SET_ERROR = 'components/globalErrors/SET_ERROR',
    REMOVE_ERROR = 'components/globalErrors/REMOVE_ERROR',
    SET_ERRORS = 'components/globalErrors/SET_ERRORS'
}

export const initialErrorsState: ErrorsState = {
    errors: []
}

export default ActionTypes
