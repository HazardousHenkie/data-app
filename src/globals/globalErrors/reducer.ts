import { ActionType } from 'typesafe-actions'
import ErrorsState from './types'

import * as actions from './actions'

import ActionTypes, { initialErrorsState } from './constants'

function errorsReducer(
    state: ErrorsState = initialErrorsState,
    action: ActionType<typeof actions>
): ErrorsState {
    switch (action.type) {
        case ActionTypes.SET_ERRORS:
            return {
                errors: action.payload
            }
        default:
            return state
    }
}

export default errorsReducer
