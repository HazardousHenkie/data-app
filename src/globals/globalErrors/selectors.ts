import { createSelector, Selector } from 'reselect'

import { ApplicationRootState } from 'types'

import { initialErrorsState } from './constants'

const makeSelectInitialErrors = (state: ApplicationRootState) =>
    state.errors || initialErrorsState

const makeSelectErrors = () =>
    createSelector(
        makeSelectInitialErrors,
        subState => subState.errors
    ) as Selector<unknown, Error[]>

export { makeSelectInitialErrors, makeSelectErrors }
