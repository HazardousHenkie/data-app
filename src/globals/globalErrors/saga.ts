import { ActionType as typeSafeAction } from 'typesafe-actions'

import { select, put } from 'redux-saga/effects'

import { initialErrorsState } from './constants'
import { makeSelectErrors } from './selectors'
import { setError, setErrors } from './actions'

export function* globalAddErrorSaga(params: typeSafeAction<typeof setError>) {
    try {
        const errors = yield select(makeSelectErrors())

        yield put(setErrors([...errors, params.payload]))
    } catch (error) {
        // eslint-disable-next-line no-console
        yield put(setError(error))
    }
}

export function* globalRemoveErrorSaga() {
    try {
        yield put(setErrors(initialErrorsState.errors))
    } catch (error) {
        yield put(setError(error))
    }
}
