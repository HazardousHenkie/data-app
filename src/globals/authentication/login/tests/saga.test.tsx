import { put } from 'redux-saga/effects'

import { ActionType as typeSafeAction } from 'typesafe-actions'
import { ResponseError } from 'utils/request'
import { setError } from 'globals/globalErrors/actions'
import ActionTypes from '../constants'

import { loginRequest, loginSuccess, loginError } from '../actions'

import loginSaga from '../saga'

describe('loginSaga Saga', () => {
    let loginSagaGenerator: Generator
    let callDescriptor: Response

    const params = {
        type: ActionTypes.LOGIN_REQUEST,
        payload: 'userId'
    }

    beforeEach(() => {
        loginSagaGenerator = loginSaga(
            params as typeSafeAction<typeof loginRequest>
        )

        callDescriptor = loginSagaGenerator.next().value
        expect(callDescriptor).toMatchSnapshot()
    })

    it('should dispatch the loginSuccess action if call was successfull', () => {
        const response = { user: { id: '', name: '' } }
        const putDescriptor = loginSagaGenerator.next(response).value

        expect(putDescriptor).toEqual(
            // eslint-disable-next-line redux-saga/no-unhandled-errors
            put(loginSuccess(response.user))
        )
    })

    it('should call the loginError and setError action if the response errors', () => {
        const response = new ResponseError(callDescriptor, 'Some error')
        const putDescriptor = loginSagaGenerator.throw(response).value
        // eslint-disable-next-line redux-saga/no-unhandled-errors
        expect(putDescriptor).toEqual(put(loginError(response)))

        const putDescriptorSecondError = loginSagaGenerator.next(response).value

        // eslint-disable-next-line redux-saga/no-unhandled-errors
        expect(putDescriptorSecondError).toEqual(put(setError(response)))
    })
})
