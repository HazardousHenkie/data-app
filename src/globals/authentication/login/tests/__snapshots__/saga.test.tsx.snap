// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loginSaga Saga should call the loginError and setError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/.netlify/functions/googleLogin",
      Object {
        "body": "{\\"authToken\\":\\"userId\\"}",
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`loginSaga Saga should dispatch the loginSuccess action if call was successfull 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "/.netlify/functions/googleLogin",
      Object {
        "body": "{\\"authToken\\":\\"userId\\"}",
        "method": "POST",
      },
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;
